feat(apple): Handle network changes reliably on macOS and iOS

jamilbk opened the pull request:
Tried to organize this PR into commits so that it's a bit easier to review.

1. Involves simplifying the logic in Adapter.swift so that us mortals can maintain it confidently:
  - The `.stoppingTunnel`, `.stoppedTunnelTemporarily`, and `.stoppingTunnelTemporarily` states have been removed.
  - I also removed the `self.` prefix from local vars when it's not necessary to use it, to be more consistent.
  - `onTunnelReady` and `getSystemDefaultResolvers` has been removed, and `onUpdateRoutes` wired up, along with cleanup necessary to support that.
2. Involves adding the `reconnect` and `set_dns` stubs in the FFI and fixing the log filter so that we can log them (see #4182 )
3. Involves getting the path update handler working well on macOS using `SystemConfiguration` to read DNS servers.
4. Involves getting the path update handler working well on iOS by employing careful trickery to prevent path update cycles by detecting if `path.gateways` has changed, and avoid setting new DNS if it hasn't.

Refs #4028 
Fixes #4297
Fixes #3565 
Fixes #3429 
Fixes #4175 
Fixes #4176 
Fixes #4309

jamilbk committed:
Greatly simplify Apple client state

jamilbk commented on swift/apple/FirezoneNetworkExtension/Adapter.swift:
```diff
@@ -465,66 +359,85 @@ extension Adapter: CallbackHandlerDelegate {
+    // sentinel, which isn't helpful to us.
+    private func resetToSystemDNSGettingBindResolvers() -> [String] {
+      var resolvers: [String] = []
+      let semaphore = DispatchSemaphore(value: 0)
```
Each path update callback spawns a thread (`workQueue.async`), so we're ok to block here with a semaphore.

jamilbk replied:
Actually, I noticed this sometimes blocked the main UI thread on iOS, so I enclosed this inside a `Task` in the calling function.

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/CallbackHandler.swift:
```diff
@@ -22,20 +22,19 @@ public protocol CallbackHandlerDelegate: AnyObject {
     tunnelAddressIPv6: String,
     dnsAddresses: [String]
   )
-  func onTunnelReady()
```
Do you now just assume that the tunnel is ready after `connect`? I think that makes sense. I guess `connlib` should fail using `on_disconnect` if we cannot initialize the device.

jamilbk replied:
We now assume it's ready after the first `onSetInterfaceConfig`

conectado replied:
https://github.com/firezone/firezone/issues/4223

thomaseizinger reviewed (commented):
Great work! I don't fully understand all the swift stuff but I have one question regarding `on_tunnel_ready`.

thomaseizinger commented on rust/connlib/clients/apple/src/lib.rs:
```diff
@@ -234,6 +214,16 @@ impl WrappedSession {
 
+    fn reconnect(&mut self) {
+        tracing::debug!("Reconnecting");
+        // self.inner.reconnect()
```
This should be wired-up before merge.

conectado commented on rust/connlib/clients/apple/src/lib.rs:
```diff
@@ -229,6 +209,16 @@ impl WrappedSession {
+
+    fn set_dns(&mut self, dns_servers: String) {
+        tracing::debug!("Setting DNS servers: {}", dns_servers);
+        // self.inner.set_dns(dns_servers)
```
We shouldn't merge until this is wired up, otherwise it will break the apple client

thomaseizinger replied:
Can we merge a minimal version of `set_dns` that allows these PRs to be independent?

I think that would be useful.

conectado commented on swift/apple/FirezoneNetworkExtension/PacketTunnelProvider.swift:
```diff
@@ -107,15 +90,15 @@ class PacketTunnelProvider: NEPacketTunnelProvider {
-      }
-    #endif
-  }
+  //  func handleTunnelShutdown(dueTo reason: TunnelShutdownEvent.Reason, errorMessage: String) {
```
Should this be commented out?

thomaseizinger reviewed (commented):
Nice clean up!

I don't really understand the Swift stuff but it looks like a great simplification :)

jamilbk commented on swift/apple/FirezoneNetworkExtension/Adapter.swift:
```diff
@@ -313,71 +226,109 @@ extension Adapter {
+        // out of a different interface even when 0.0.0.0 is used as the source.
+        // If our primary interface changes, we can be certain the old socket shouldn't be
+        // used anymore.
+        if (path.availableInterfaces.first?.name != primaryInterfaceName) {
```
@thomaseizinger @conectado I only call reconnect if the primary interface changes.

thomaseizinger replied:
Great!

jamilbk commented on swift/apple/FirezoneNetworkExtension/Adapter.swift:
```diff
@@ -313,71 +226,109 @@ extension Adapter {
+            let resolvers = getSystemDefaultResolvers(interfaceName: path.availableInterfaces.first?.name)
 
-      self.logger.log("Adapter.didReceivePathUpdate: Back online. Starting connlib.")
+            if lastFetchedResolvers.sorted() != resolvers.sorted() {
```
I only call set_dns if the resolvers have changed now.

jamilbk commented on swift/apple/FirezoneNetworkExtension/Adapter.swift:
```diff
@@ -313,71 +226,109 @@ extension Adapter {
   }
+
+  #if os(iOS)
+    private func shouldFetchSystemResolvers(path: Network.NWPath) -> Bool {
```
Caveat, we only fetch new system resolvers if the available gateways for this path has changed. It's not perfect but gets the job done.

jamilbk commented on swift/apple/FirezoneNetworkExtension/Adapter.swift:
```diff
@@ -389,56 +340,26 @@ extension Adapter: CallbackHandlerDelegate {
+        networkSettings.apply()
+      case .stoppedTunnel:
+        logger.error(
+          "\(#function): Unexpected state: \(self.state)")
```
We could probably even collapse these states further into just `.startedTunnel` and `.stoppedTunnel`. There's not much that can go wrong between `.startingTunnel` and `.tunnelReady`.

thomaseizinger replied:
Do you have a timeout for going into `tunnelReady`? If not then the state is already pointless I'd say.

All of these states should be able to be removed. Either you have a pointer or you don't. `connect` gives you the pointer, `disconnect` "consumes" it. `on_disconnected` you should throw it away.

Your state machine should just be an `Option` or whatever the equivalent of that is in Swift :)

jamilbk replied:
`on_disconnected` is used to notify the client their session has ended (portal 401), but otherwise I'm with you.

We have to call a completionHandler when the tunnel "is up" but that could just be right when first `onSetInterfaceConfig` happens, like we do now.

thomaseizinger replied:
> `on_disconnected` is used to notify the client their session has ended (portal 401)

It is also used when connlib crashes. I think the main semantics here is, once you receive `on_disconnected`, the pointer you had to connlib is invalid and should no longer be used. Why it was disconnected is in the error which at the moment is unfortunately untyped.

thomaseizinger replied:
> We have to call a completionHandler when the tunnel "is up" but that could just be right when first `onSetInterfaceConfig` happens, like we do now.

Who mandates this completion handler? Could you call it right after you initialized connlib using `connect`?

jamilbk replied:
> Who mandates this completion handler? Could you call it right after you initialized connlib using connect?

We could, but we need interface IPs at a minimum to bring the interface up, so that's why I wait for the `onSetInterfaceConfig`.

Once the completionHandler is called, Apple considers the interface active and starts sending packets to it.

jamilbk replied:
> Do you have a timeout for going into tunnelReady? If not then the state is already pointless I'd say.

Not on the Apple side -- I think it just relies on connlib's callback. Yeah let me take another stab at cleaning this up some more! :D

jamilbk replied:
Done, removed the pointless `.startingTunnel` state.

jamilbk replied:
https://github.com/firezone/firezone/pull/4133/commits/bd4d248c547ba15eea071189317c33cf939db9b5

jamilbk commented on swift/apple/FirezoneNetworkExtension/Adapter.swift:
```diff
@@ -447,81 +368,106 @@ extension Adapter: CallbackHandlerDelegate {
+
+      // HACK: Define more connlib error types across the FFI so we can switch on them
+      // directly and not just sign the user out every time this is called.
+      if error.contains("401 Unauthorized") {
```
FYI I now explicitly check for a 401 Unauthorized before clearing the token. Otherwise it just signs the user out quietly.

thomaseizinger replied:
I guess we can refactor this to a typed API once we move to UniFFI!

ReactorScram replied:
does this matter for other Clients? I don't think I check for 401 anywhere.

jamilbk committed:
Refactor how we store, parse, and display resources

jamilbk committed:
Check for tunnel status moving back to .invalid on iOS

jamilbk committed:
Stricter settings field validators

jamilbk committed:
Fix spelling

jamilbk committed:
Don't allow clicking Sign out when disconnecting

jamilbk committed:
Set DNS immediately after WrappedSession.connect

jamilbk committed:
Use kSecAttrAccessGroup for shared Keychain access on macOS too

jamilbk committed:
Simplify signed out state

jamilbk commented on swift/apple/FirezoneKit/Sources/FirezoneKit/Helpers/AppLogger.swift:
```diff
@@ -9,446 +9,128 @@ import Foundation
-
-    // Look for an existing log file with this log index
-    var existingLogFiles: [URL] = []
-    if let enumerator = fileManager.enumerator(
```
@ReactorScram @conectado I believe this may have the thing triggering the App Store privacy notification.

jamilbk committed:
rework AppLogger

ReactorScram commented on swift/apple/FirezoneNetworkExtension/NetworkSettings.swift:
```diff
@@ -9,108 +9,56 @@ import NetworkExtension
-  public var routes6: [NEIPv6Route] = []
+  // Constants
+  private weak var packetTunnelProvider: NEPacketTunnelProvider?
+  private(set) var logger: AppLogger
```
These are not constant in the sense of Rust's `const`, but they're objects that won't be replaced during the lifetime of this `NetworkSettings` object?

jamilbk replied:
Correct, that's probably not a good comment. I'll fix it.

ReactorScram commented on swift/apple/README.md:
````diff
@@ -110,3 +110,12 @@ cd swift/apple
 rm -rf $HOME/Library/Group\ Containers/47R2M6779T.group.dev.firezone.firezone/Library/Caches/logs/connlib
 ```
+
+### Clearing the Keychain item
+
+Sometimes it's helpful to be able to test how the app behaves when the keychain
+item is missing. You can remove the keychain item with the following command:
+
+```bash
+security delete-generic-password -s "dev.firezone.firezone"
+```
````
👍

ReactorScram commented on rust/connlib/clients/apple/src/lib.rs:
```diff
@@ -44,6 +44,9 @@ mod ffi {
             callback_handler: CallbackHandler,
         ) -> Result<WrappedSession, String>;
 
+        fn reconnect(&mut self);
+        #[swift_bridge(swift_name = "setDns")]
+        fn set_dns(&mut self, dns_servers: String);
```
I'll leave some redundant questions just to make sure I understand how the Apple Client fits together.

So this part is exposing the Rust functions to Swift, we're declaring these two new functions. This is still on the Rust side, so I assume it would error if the signature is wrong.

jamilbk replied:
Yeah! The macros in here are used to generate the bindings.

ReactorScram commented on swift/apple/Firezone/Application/FirezoneApp.swift:
```diff
@@ -74,31 +75,24 @@ struct FirezoneApp: App {
-        }
-      }
-    }
+    public var appStore: AppStore?
```
"store" here means some persistent variable storage, it's not related to the Apple App Store package manager, right?

jamilbk replied:
Yeah, I was confused by that too at first. That's correct.

It could probably use a rename.

ReactorScram commented on swift/apple/FirezoneKit/Sources/FirezoneKit/Features/AskPermissionView.swift:
```diff
@@ -33,14 +33,13 @@ public final class AskPermissionViewModel: ObservableObject {
-      .sink { [weak self] tunnelAuthStatus in
+    tunnelStore.$status
+      .sink { [weak self] status in
         guard let self = self else { return }
```
This is equivalent to the `let-else` we use in Rust? The `else` is required to diverge, and it doesn't do anything special, it's just a more idiomatic way to do an early return?

jamilbk replied:
That's correct. Guards are pretty heavily used in Swift and they turn `Optional(obj)` into just `obj` so you don't need the optional modifier in the remainder of the block body.

ReactorScram commented on swift/apple/FirezoneKit/Sources/FirezoneKit/Features/MainView.swift:
```diff
@@ -15,70 +15,47 @@ import SwiftUI
+      tunnelStore.$status
         .receive(on: mainQueue)
         .sink { [weak self] status in
-          self?.tunnelStatus = status
+          guard let self = self else { return }
           if status == .connected {
-            self?.appStore.tunnelStore.beginUpdatingResources()
+            self.tunnelStore.beginUpdatingResources()
           } else {
-            self?.appStore.tunnelStore.endUpdatingResources()
+            self.tunnelStore.endUpdatingResources()
           }
         }
         .store(in: &cancellables)
```
This was hard to understand, so I asked ChatGPT and tried to look it up in the Apple docs (https://developer.apple.com/documentation/combine/published) and this is what I heard:

- The `$` sigil means `status` is a Published property that can be subscribed to.
- `receive` means we are subscribing to it
- ChatGPT says `mainQueue` refers to a process-wide main thread (interchangeably called a 'queue') for processing UI events. I've seen the same pattern with Qt's signals and slots, and C# / .NET's `Dispatcher` class, so this is just saying, "Here's an event, make sure it's processed on the main thread so we can touch the UI"
- `sink` is the closure / callback that we're adding as a subscriber
- `weak self` is a weak pointer to avoid reference cycles (had that exact problem in C# with event handlers), so if our `MainViewModel` needs to be freed, it can be freed, and the subscriber closure will just bail out when it sees that the weak pointer has become null.
- `guard let` is required to dereference the weak pointer, hopefully this is a compile-time error in Swift. It must be similar to the `Weak` Arc in Rust that can upgrade to an `Option<Arc>`.
- `beginUpdatingResources` and `endUpdatingResources` I assume control an async task for requesting the resources from the tunnel process. "begin" probably does nothing if the task is already running, and "end" tries to cancel the task, and does nothing if it's not running.
- ChatGPT told me that the main queue will drop old events or use backpressure to avoid using an unbounded amount of RAM if the subscribers can't process events fast enough. I'm not sure if both or either of those things are true?
- ChatGPT said that setting the property always calls the subscriber, even if the value didn't change. (Makes sense, some values might not be equatable)
- `store(in: &cancellables)` means we're getting back a handle for this subscription and storing it in a vector or set or something. ChatGPT tells me that the `&` sigil here means it's an in-out parameter, so we're modifying this collection. And we would use that later when we're closing the app, to cancel all subscriptions.

It's very different from Rust. Did I understand it correctly?

jamilbk replied:
> It's very different from Rust. Did I understand it correctly?

Spot on! I just learned a couple things from this too.

> ChatGPT told me that the main queue will drop old events or use backpressure to avoid using an unbounded amount of RAM if the subscribers can't process events fast enough. I'm not sure if both or either of those things are true?

Probably true. I would trust ChatGPT's "intuition" on this, because it's one of the most common concepts in Swift UI programming.

ReactorScram commented on swift/apple/FirezoneKit/Sources/FirezoneKit/Features/MainView.swift:
```diff
@@ -15,70 +15,47 @@ import SwiftUI
         .receive(on: mainQueue)
-        .sink { [weak self] resources in
-          guard let self = self else { return }
-          self.orderedResources = resources.orderedResources.map {
```
Are the resources not ordered anymore? Doesn't connlib alphabetize them?

jamilbk replied:
They are, but this was named this because there were two types in the Apple client `resources` and `orderedResources`, before connlib sorted them.

The `resources` was removed, leaving `orderedResources`, which to me implies we ordered them somewhere, but we don't. We don't actually care what order they're in anymore. We populate the list from the initial `resourceListJSON` and then from there we update the items in-place using their `address` and `name`.

ReactorScram commented on swift/apple/FirezoneKit/Sources/FirezoneKit/Features/MainView.swift:
```diff
@@ -90,56 +67,53 @@ import SwiftUI
       List {
         Section(header: Text("Authentication")) {
```
Where does this "Authentication" text go? Is it used for accessibility or something?

jamilbk replied:
This is used on iOS as a Section header when Signed in.

![Screenshot 2024-03-26 at 7 24 04 PM](https://github.com/firezone/firezone/assets/167144/589f7c9f-5d26-46d1-a7c4-2d7adfe23171)

ReactorScram commented on swift/apple/FirezoneKit/Sources/FirezoneKit/Features/SettingsView.swift:
```diff
@@ -16,71 +16,51 @@ enum SettingsViewError: Error {
-        .sink { [weak self] tunnelAuthStatus in
+        .sink { [weak self] settings in
           guard let self = self else { return }
-          self.advancedSettings =
-            authStore.tunnelStore.advancedSettings() ?? AdvancedSettings.defaultValue
```
Does the Apple client save the default settings to disk (or to 'store' I guess?) so that there's always a valid settings file?

I was wondering this, I think for the Tauri client I don't save it, which means if we change the default settings in an update, it will change for anyone who hasn't hit "Save" ever before, but not for anyone who has ever changed the settings. I'm not sure which way makes more sense.

jamilbk replied:
The persistence layer for the Apple client tunnel settings is actually the `NETunnelProviderManager` `manager` object's `protocolConfiguration`. Or the thing you see in System Prefs.

This is the only thing (besides the filesystem and keychain) we can read from the root Tunnel process.

The `serverAddress` field is required, which we use to store the API URL to the portal, so we need it to be populated anyway. So we do store the default settings there when creating the Manager.

ReactorScram commented on swift/apple/FirezoneKit/Sources/FirezoneKit/Features/SettingsView.swift:
```diff
@@ -235,7 +206,7 @@ public struct SettingsView: View {
       markdown: """
         **WARNING:** These settings are intended for internal debug purposes **only**. \
-        Changing these is not supported and will disrupt access to your Firezone resources.
+        Changing these will disrupt access to your Firezone resources.
```
Need to change the corresponding string in the Tauri Client. It'd be cool if we had a string table shared across all Clients.

ReactorScram commented on swift/apple/FirezoneKit/Sources/FirezoneKit/Features/SettingsView.swift:
```diff
@@ -347,26 +318,26 @@ public struct SettingsView: View {
               text: Binding(
-                get: { model.advancedSettings.apiURLString },
-                set: { model.advancedSettings.apiURLString = $0 }
+                get: { model.settings.apiURL },
```
Tauri Client still calls them "Advanced settings", might refactor that after GA.

jamilbk replied:
Yeah had that thought too.

ReactorScram commented on swift/apple/FirezoneKit/Sources/FirezoneKit/Helpers/SharedAccess.swift:
```diff
@@ -69,11 +69,11 @@ public struct SharedAccess {
-  public static var tunnelShutdownEventFileURL: URL {
-    baseFolderURL.appendingPathComponent("tunnel_shutdown_event_data.json")
+  public static var providerStopReasonURL: URL {
+    baseFolderURL.appendingPathComponent("reason")
```
No file extension?

jamilbk replied:
Its raw value is an `Int` (I think `4` in the onDisconnect auth canceled case)

ReactorScram commented on swift/apple/FirezoneKit/Sources/FirezoneKit/Keychain/Keychain.swift:
```diff
@@ -18,23 +18,22 @@ public enum KeychainError: Error {
+  private let label = "Firezone token"
+  private let description = "Firezone access token used to authenticate the client."
+  private let account = "Firezone"
+  private let service = Bundle.main.bundleIdentifier!
```
Had to look this up, ChatGPT says `!` is like `unwrap()`. I think I saw this in JS or C# but it was just surprising to see it in a variable declaration.

jamilbk replied:
correct, it will "panic"

ReactorScram commented on swift/apple/FirezoneKit/Sources/FirezoneKit/Keychain/Keychain.swift:
```diff
@@ -18,23 +18,22 @@ public enum KeychainError: Error {
+  private let account = "Firezone"
+  private let service = Bundle.main.bundleIdentifier!
+
   private let workQueue = DispatchQueue(label: "FirezoneKeychainWorkQueue")
```
So a `DispatchQueue` is sort of like a message queue for events, it's not quite a thread but it probably runs on a thread pool somewhere, and the main queue is equivalent to the main thread in other UI frameworks that have thread restrictions?

jamilbk replied:
That's correct. It's mostly used in the Apple client as an ordered async thread queue.

ReactorScram commented on swift/apple/FirezoneKit/Sources/FirezoneKit/Keychain/Keychain.swift:
```diff
@@ -48,128 +47,73 @@ public actor Keychain {
-      let query =
-        [
+  func add(token: Token) async throws -> PersistentRef {
+    return try await withCheckedThrowingContinuation { [weak self] continuation in
```
ChatGPT says this has something to do with inter-operating between callbacks and async `Task`s.

I see a closure down here, and that calls `continuation.resume`, so I guess we're pausing this Task until the workQueue does something in that closure?

jamilbk replied:
Yep! Continuations are how you can exit from an async context back to the caller, almost like a `yield`

ReactorScram commented on swift/apple/FirezoneKit/Sources/FirezoneKit/Keychain/Keychain.swift:
```diff
@@ -48,128 +47,73 @@ public actor Keychain {
-        return access
-      } else {
-        throw KeychainError.nilResultFromAppleSecCall(call: "SecAccessCreate")
+        continuation.resume()
```
So this means, if the closure succeeds, we resume `func update` and it returns success.

jamilbk replied:
Well technically returns `Void`, but yeah it's successful.

ReactorScram commented on swift/apple/FirezoneKit/Sources/FirezoneKit/Keychain/KeychainStorage.swift:
```diff
@@ -8,45 +8,37 @@ import Dependencies
-  var delete: @Sendable (Keychain.PersistentRef) async throws -> Void
-  var loadAttributes: @Sendable (Keychain.PersistentRef) async -> Keychain.TokenAttributes?
-  var searchByAuthBaseURL: @Sendable (URL) async -> Keychain.PersistentRef?
+  var add: @Sendable (Keychain.Token) async throws -> Keychain.PersistentRef
+  var update: @Sendable (Keychain.Token) async throws -> Void
+  var search: @Sendable () async -> Keychain.PersistentRef?
```
ChatGPT says these are variables storing functions? Like function pointers? How does that work?

jamilbk replied:
These are just closures defined as instance variables on the struct I believe.

On L21-L23 you can see how they're used. I think the original author of this class had started writing tests for this but didn't finish...

ReactorScram commented on swift/apple/FirezoneKit/Sources/FirezoneKit/Keychain/KeychainStorage.swift:
```diff
@@ -8,45 +8,37 @@ import Dependencies
-      searchByAuthBaseURL: { _ in
-        nil
+      search: {
+        return UUID().uuidString.data(using: .utf8)!
```
I didn't see where `UUID()` comes from. Is that a compile-time const UUID for Firezone's keychain stuff, like the bundle ID? It's not generating a random one at runtime, I assume?

jamilbk replied:
`import Foundation` above.
Kinda like Swift's stdlib

ReactorScram commented on swift/apple/FirezoneKit/Sources/FirezoneKit/Models/Settings.swift:
```diff
@@ -6,49 +6,73 @@
 
-struct AdvancedSettings: Equatable {
-  var authBaseURLString: String {
-    didSet { if oldValue != authBaseURLString { isSavedToDisk = false } }
```
Is `oldValue` really a keyword in Swift?

jamilbk replied:
Lol, looks like it.

It might be syntax sugar from `Foundation`. It's a block variable from the `didSet` function probably.

ReactorScram commented on swift/apple/FirezoneKit/Sources/FirezoneKit/Views/MenuBar.swift:
```diff
@@ -32,60 +33,63 @@
+
+          let newResources = decodeResources(json)
+
+          // Update menu in-place for smooth sexy easy on the eyes UI updates
```
```suggestion
          // Update menu in-place for smooth, silky, easy-on-the-eyes UI updates
```

Better safe than sorry.

So does that mean on macOS it can update the menu with an animation even while it's open?

jamilbk replied:
Lol, sorry, you're right. I wrote that late at night 🙈

ReactorScram commented on swift/apple/FirezoneKit/Sources/FirezoneKit/Views/MenuBar.swift:
```diff
@@ -32,60 +33,63 @@
+      return try? JSONDecoder().decode([Resource].self, from: data)
+    }
+
+    // FIXME: Use SwiftUI for the menubar
```
For after GA?

jamilbk replied:
Yeah, there's still lots of clean up todo. Iterations.

ReactorScram commented on swift/apple/FirezoneKit/Sources/FirezoneKit/Views/MenuBar.swift:
```diff
@@ -371,38 +360,40 @@
+      if isMenuVisible && tunnelStore.status == .connected {
+        tunnelStore.beginUpdatingResources()
+      } else {
+        tunnelStore.endUpdatingResources()
```
This part I don't quite get. Why do we have to begin updating resources? Does the tunnel process not push a list of resources to the GUI process when the connlib callback fires?

Also this looks familiar, I feel like another piece of code does the same thing. Either that or I'm remembering when I saw this piece in a commit diff this morning. Maybe one is for iOS and one is for macOS?

jamilbk replied:
Yeah the Menubar and Mainview are similar, but slightly different. It could probably be cleaned up with smarter SwiftUI code that handles both platforms.

The issue is that the tunnel process _can't_ push anything to the GUI process. There's no IPC mechanism for that, unless you get creative with daemons and stuff.

So `beginUpdatingResources` is a timer that's instantiated when the list of resources is visible and checks every 1s using a `sha256` of the list to determine if it needs to be sent over the IPC boundary.

ReactorScram commented on swift/apple/FirezoneKit/Sources/FirezoneKit/Views/MenuBar.swift:
```diff
@@ -371,38 +360,40 @@
+          let menuItem = createResourceMenuItem(title: element.name, submenuTitle: element.address)
+          menu.insertItem(menuItem, at: index + offset)
         case .remove(let offset, element: _, associatedWith: _):
-          menu.removeItem(at: baseIndex + offset)
-          orderedResources.remove(at: offset)
+          menu.removeItem(at: index + offset)
```
Hm, the Apple client does a diff and mutates the menu instead of blowing it up and regenerating it all? Hope it doesn't de-sync.

jamilbk replied:
Yeah it's been pretty stable so far. I think this can be redone though as part of #3514

ReactorScram commented on swift/apple/FirezoneNetworkExtension/DeviceMetadata.swift:
```diff
@@ -0,0 +1,57 @@
+//
+//  DeviceMetadata.swift
+//  Firezone
+//
+//  Created by Jamil Bou Kheir on 2/23/24.
+//
```
Does XCode do this? Cause Git already tracks all that info, plus it doesn't use mm/dd/yy

jamilbk replied:
Yeah, Xcode does that. We probably need a consistent header.

ReactorScram commented on swift/apple/FirezoneNetworkExtension/DeviceMetadata.swift:
```diff
@@ -0,0 +1,57 @@
+  // if that doesn't exist. The Firezone ID is a UUIDv4 that is used to dedup this device
+  // for upsert and identification in the admin portal.
+  public static func getOrCreateFirezoneId(logger: AppLogger) -> String {
+    let fileURL = SharedAccess.baseFolderURL.appendingPathComponent("firezone-id")
```
No `.txt` or anything?

jamilbk replied:
These files aren't really accessible outside Firezone. I guess you could append extensions but extensions aren't really common for random app-specific files on Unix I think.

ReactorScram commented on swift/apple/FirezoneNetworkExtension/DeviceMetadata.swift:
```diff
@@ -0,0 +1,57 @@
+      let newUUIDString = UUID().uuidString
+
+      do {
+        try newUUIDString.write(to: fileURL, atomically: true, encoding: .utf8)
```
Apple just has a flag for atomic file writes? I'm jealous :P

jamilbk replied:
Yeah there are some nice things actually in the Apple Garden™

ReactorScram commented on swift/apple/FirezoneNetworkExtension/NetworkSettings.swift:
```diff
@@ -9,108 +9,56 @@ import NetworkExtension
-    }
-    self.hasUnappliedChanges = true
-  }
+  public var tunnelAddressIPv4: String?
+  public var tunnelAddressIPv6: String?
+  public var dnsAddresses: [String] = []
```
Is there a reason these are stored as String even though the routes have a more concrete type?

jamilbk replied:
Yeah, Apple's API expects them as Strings, so there's no point going further than that. They're sent from connlib and assumed to be valid.

ReactorScram commented on swift/apple/FirezoneNetworkExtension/PacketTunnelProvider.swift:
```diff
@@ -86,35 +63,42 @@ class PacketTunnelProvider: NEPacketTunnelProvider {
+        // Remove the passwordReference from our configuration so that it's not used again
+        // if the app is re-launched. There's no good way to send data like this from the
+        // Network Extension to the GUI, so save it to a file for the GUI to read later.
+        try String(reason.rawValue).write(to: SharedAccess.providerStopReasonURL, atomically: true, encoding: .utf8)
```
So with using files as IPC...

- There's no chance the GUI will read this file before the tunnel gets a chance to write it?
- There's no chance the GUI would crash and then read the file after it's stale and mis-understand what state we're in?
- Does the GUI delete the file when it reads it, to avoid processing the same event twice?

jamilbk replied:
- Not really, the sequence of events is:
  - Tunnel process writes reason to file
  - Tunnel process exits
  - GUI receives `NEVPNStatusDidChange`
  - GUI reads the file and acts accordingly, deleting it
- Hm, it could, but this file is solely for (at this point) showing a UI notification. It's a boolean flag that tells it "You were unexpectedly signed out, sign in again". If the GUI crashes before it's able to "consume" the notification, it'll just show it again on next launch, no biggie (other than an annoyance to the user).
- Yes

ReactorScram commented on swift/apple/FirezoneNetworkExtension/PacketTunnelProvider.swift:
```diff
@@ -127,7 +111,7 @@ extension NEProviderStopReason: CustomStringConvertible {
     case .unrecoverableNetworkChange: return "Unrecoverable network change"
     case .providerDisabled: return "Provider disabled"
-    case .authenticationCanceled: return "Authentication cancelled"
+    case .authenticationCanceled: return "Authentication canceled"
```
A few comments in the Rust code spelled this wrong too, it's one of the only words I always guess wrong for spelling.

thomaseizinger replied:
It is "cancelled" in British-English :)

ReactorScram commented on swift/apple/FirezoneNetworkExtension/SystemConfigurationResolvers.swift:
```diff
@@ -0,0 +1,124 @@
+
+class SystemConfigurationResolvers {
+  private let logger: AppLogger
+  private var dynamicStore: SCDynamicStore?
```
I looked this up and the Apple docs just said it's a handle to "the dynamic store of a running system". It didn't say what a dynamic store is.

I'm guessing it's a key-value store that doesn't survive reboots?

jamilbk replied:
It's actually the system's configuration and is not app specific. That's why this API is not available for iOS. I don't believe it's writable due to App sandboxing.

ReactorScram commented on swift/apple/FirezoneNetworkExtension/SystemConfigurationResolvers.swift:
```diff
@@ -0,0 +1,124 @@
+  private var dynamicStore: SCDynamicStore?
+
+  // Arbitrary name for the connection to the store
+  private let storeName = "dev.firezone.firezone.dns" as CFString
```
If we have to provide a name, does that means the system doesn't check our bundle ID or process ID for sandboxing? Could another app write things into our dynamic store?

jamilbk replied:
Nah, I _think_ we might be able to write to the SC store, but I'm not sure, since our tunnel process is supposed to be sandboxed (may need an entitlement for that).

https://developer.apple.com/documentation/systemconfiguration/scpreferencessetspecific

ReactorScram commented on swift/apple/FirezoneNetworkExtension/Adapter.swift:
```diff
@@ -7,48 +8,23 @@ import Foundation
-    case .tunnelReady: return "tunnelReady"
-    case .stoppingTunnel: return "stoppingTunnel"
-    case .stoppedTunnel: return "stoppedTunnel"
-    case .stoppingTunnelTemporarily: return "stoppingTunnelTemporarily"
-    case .stoppedTunnelTemporarily: return "stoppedTunnelTemporarily"
+    case .tunnelStarted: return "tunnelStarted"
+    case .tunnelStopped: return "tunnelStopped"
     }
   }
```
No Swift equivalent of `#[derive(Debug, Display)]`?

ReactorScram commented on swift/apple/FirezoneNetworkExtension/Adapter.swift:
```diff
@@ -72,7 +47,19 @@ class Adapter {
+    private var gateways: [Network.NWEndpoint] = []
+  #endif
+
+  /// Track our last fetched DNS resolvers to know whether to tell connlib they've updated
+  private var lastFetchedResolvers: [String] = []
```
I think connlib already does that on main, although it doesn't filter out the sentinels correctly yet. Gabi and I each have different PRs open for that.

ReactorScram commented on swift/apple/FirezoneNetworkExtension/Adapter.swift:
```diff
@@ -72,7 +47,19 @@ class Adapter {
+
+  /// Private queue used to ensure consistent ordering among path update and connlib callbacks
+  /// This is the primary async primitive used in this class.
   private let workQueue = DispatchQueue(label: "FirezoneAdapterWorkQueue")
```
These are scoped to the application or the process, right? I do see the Apple docs saying e.g. `com.example.myqueue` but hopefully no other app sees these queues.

ReactorScram commented on swift/apple/FirezoneNetworkExtension/Adapter.swift:
```diff
@@ -83,227 +70,145 @@ class Adapter {
-          onStarted: completionHandler
+    self.logger.log("Adapter.start: Starting connlib")
+    do {
+      // Grab a session pointer
```
```suggestion
      // Create a session
```

We're not getting a handle to an existing Session, right?

ReactorScram commented on swift/apple/FirezoneNetworkExtension/Adapter.swift:
```diff
@@ -83,227 +70,145 @@ class Adapter {
+      self.state = .tunnelStarted(session: session)
+
+      // Start listening for network change events. The first few will be our
+      // tunnel interface coming up, but that's ok -- it will trigger a `set_dns`
+      // connlib.
```
```suggestion
      // tunnel interface coming up, but that's ok -- connlib will behave correctly anyway
```

Also, does this mean there's a gap where network events could be accidentally dropped during startup? Does that matter?

ReactorScram commented on swift/apple/FirezoneNetworkExtension/Adapter.swift:
```diff
@@ -83,227 +70,145 @@ class Adapter {
+
+      // Tell the system the tunnel is up, moving the tunnelManager status to
+      // `connected`.
+      completionHandler(nil)
```
Completion handlers are async callbacks, like the Promises in JS?

ReactorScram commented on swift/apple/FirezoneNetworkExtension/Adapter.swift:
```diff
@@ -83,227 +70,145 @@ class Adapter {
       guard let self = self else { return }
 
-      if referenceVersionString == self.displayableResources.versionString {
+      if hash == Data(SHA256.hash(data: Data((resourceListJSON ?? "").utf8))) {
```
Why wouldn't string equality work here?

ReactorScram commented on swift/apple/FirezoneNetworkExtension/Adapter.swift:
```diff
@@ -313,71 +218,110 @@ extension Adapter {
+      }
+
+      if shouldFetchSystemResolvers(path: path) {
+        // Spawn a new thread to avoid blocking the UI on iOS
```
No chance that thread / task can leak?

ReactorScram commented on swift/apple/FirezoneKit/Sources/FirezoneKit/Helpers/AppLogger.swift:
```diff
@@ -9,446 +9,128 @@ import Foundation
-  private let logger: Logger
-
   // All log writes happen in the workQueue
   private let workQueue: DispatchQueue
```
Does this do backpressure if the queue gets too long?

ReactorScram reviewed (approved):
I left a lot of questions relating to trying to learn Swift.

This was hard to review. I see a lot of stuff changed names, some diffs look like just formatting changes too. So I may have overlooked semantic changes, in addition to the expected amount of not understanding it all yet.

I'd rather have formatting and renaming separated from semantic changes in general.

The control flow also seems confusing to me, but maybe that's something we could clean up after GA. Maybe the flow in the Tauri Client only makes sense to me because I wrote it.

jamilbk committed:
fix(apple): Don't crash on tunnel save if the configuration has been removed (#4324)

- Handles an edge case where we would crash if the tunnel configuration
was removed in the middle of saving Settings.
- Handles an edge case where our tunnel configuration may have been
changed by the system -- if it was disabled, we disconnect.

jamilbk committed:
fix(apple): Fix stuck menubar icon state (#4323)

When binding to published variables from other objects, we must use the
parameter received in the callback to act upon, and not re-read the
stored parameter on the associated object again.

The stored parameter can be stale, which leads to UI bugs where we
transitioned state but the menubar icon or menu did not get updated.

<img width="274" alt="Screenshot 2024-03-25 at 11 45 35 PM"
src="https://github.com/firezone/firezone/assets/167144/17a26b52-0599-4523-a7d1-d1bb2e617c44">


This fixes a long-standing bug that has occurred off and on since the
early versions.

jamilbk committed:
Update swift/apple/FirezoneKit/Sources/FirezoneKit/Views/MenuBar.swift

Co-authored-by: Reactor Scram <ReactorScram@users.noreply.github.com>
Signed-off-by: Jamil <jamilbk@users.noreply.github.com>

jamilbk commented on swift/apple/FirezoneNetworkExtension/NetworkSettings.swift:
```diff
@@ -9,108 +9,56 @@ import NetworkExtension
 
-  public var routes4: [NEIPv4Route] = []
-  public var routes6: [NEIPv6Route] = []
+  // Constants
```
```suggestion
  // These will only be initialized once and then don't change
```

jamilbk committed:
Update swift/apple/FirezoneNetworkExtension/NetworkSettings.swift

Signed-off-by: Jamil <jamilbk@users.noreply.github.com>
